### PR TITLE
LCOW: Dedicated scratch space for service VM utilities

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

Adds a dedicated space for the service VM utilities to use as a scratch space. 

@johnstep PTAL. 
@gupta-ak FYI

### CODE REVIEW

File: daemon/graphdriver/lcow/lcow.go [hunk 1]
```diff
@@ -51,8 +55,13 @@ type Driver struct {
 	cachedSandboxFile string
 	// options are the graphdriver options we are initialised with
 	options []string
-	// JJH LIFETIME TODO - Remove this and move up to daemon. For now, a global service utility-VM
+	// config is the representation of the SVM.
+	// @jhowardmsft LIFETIME TODO - For now, a global service utility-VM
 	config client.Config
+	// svmScratchSpaceFile is a host location for a dedicate scratch space
```
Comment (johnstep) [inline daemon/graphdriver/lcow/lcow.go]: dedicated

File: daemon/graphdriver/lcow/lcow.go [hunk 2]
```diff
@@ -102,10 +117,36 @@ func (d *Driver) startUvm(context string) error {
...
 	}
+
+	// Hot-add the scratch-space if not already attached
+	if !scratchAttached {
+		logrus.Debugf("lcowdriver: startuvm: (%s) creating an SVM scratch", context)
+		if err := d.config.CreateSandbox(d.svmScratchSpaceFile, client.DefaultSandboxSizeMB, d.cachedSandboxFile); err != nil {
+			return fmt.Errorf("failed to create SVM scratch VHDX (%s): %s", context, err)
+		}
+		logrus.Debugf("lcowdriver: startuvm: (%s) hot-adding an SVM scratch", context)
+		if err := d.config.HotAddVhd(d.svmScratchSpaceFile, `/mnt/gcs/LinuxServiceVM/scratch`); err != nil {
```
Comment (johnstep) [inline daemon/graphdriver/lcow/lcow.go]: Maybe define a constant for the container mount path, to avoid duplication.

**Review by johnstep — APPROVED** [review]
LGTM

File: daemon/graphdriver/lcow/lcow.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Agreed. However this code will change shortly to have a single service VM. This is just a single step forward.

Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Single per container. Just to clarify

**Comment by jhowardmsft** [comment]
Rebased on master.

**Comment by jhowardmsft** [comment]
Green finally. Power CI must have run a dozen times - very flaky at the moment 😢

**Comment by jhowardmsft** [comment]
@thaJeztah @vdemeester PTAL if you have time. Thanks :smile:

File: daemon/graphdriver/lcow/lcow.go [hunk 3]
```diff
@@ -102,10 +119,36 @@ func (d *Driver) startUvm(context string) error {
 		return fmt.Errorf("failed to generate default gogcs configuration (%s): %s", context, err)
 	}
 
+	scratchAttached := false
+	if _, err := os.Stat(d.svmScratchSpaceFile); err == nil {
+		// We have a scratch space already, so just attach it as a mapped virtual disk
+		logrus.Debugf("lcowdriver: startuvm: (%s) attaching pre-existing scratch", context)
```
Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: not introduced by this PR, but `context` is a confusing name (as it's not a [`context`](https://blog.golang.org/context))

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**